### PR TITLE
Massive Manifest and Recipe related types/nulls cleanup

### PR DESCRIPTION
**Pull request by lindner**
- Add types for local variables in many places.
- Switch some nullable variable to just undefined for ease of use
  - example speculator in planner.ts
- For concat'ing results add the type to the bare array
  - [].concat becomes ([] as FooType).concat
- Fix overlapping variable results/suggestionResults in planner.ts
- Add MatchingHandle type in recipe-index.ts
- Add strict Loader usage in manifest.ts
  - Switch planner-tests to use a subclass of StubLoader
- Replace some asserts with throw Error to help Typescript narrow variables.
- Add ManifestParseOptions and ManifestLoadOptions types in manifest.ts, remove unused options
- Noted some areas where typing fails as TODOs
- Fixed an issue with incorrect syntax for HandleConnections: param vs name
- Removed more incorrect calls to Manifest.parse that did not use options properly (firebase-tests.ts)
- Added implements Comparable to classes that implement it.
- used CloneMap in recipes more consistently
- Fixed issues where setter/getter were mismatched
- Use ToStringOptions and other option classes consistently in recipes code.
- Switched some recipe toString() args because of type signature mismatch in child classes.
- Avoided instanceof checking in findSlotById, use subclass implementation.
- use checkNotNull/checkDefined in more tests to avoid nullable issues.

### CODE REVIEW

**Comment by lindner** [comment]
[LMK if you think I should break this down into smaller set of changes]

File: src/planning/planner.ts [hunk 1]
```diff
@@ -206,9 +206,9 @@ export class Planner implements InspectablePlanner {
       }
       return results;
     })));
-    results = [].concat(...results);
+    const suggestionResults = ([] as Suggestion[]).concat(...results);
```
Comment (lindner) [inline src/planning/planner.ts]: variable reuse

File: src/runtime/manifest.ts [hunk 2]
```diff
@@ -1241,12 +1255,16 @@ ${e.message}
     } else {
       model = entities.map(value => ({id: value.id, value}));
     }
-    await store.fromLiteral({version, model});
+
+    // TODO(lindner): fromLiteral is not declared in the StorageProviderBase/StorageStub interface
+    // tslint:disable-next-line: no-any
+    await (store as any).fromLiteral({version, model});
```
Comment (lindner) [inline src/runtime/manifest.ts]: @shans should we add fromLiteral to the public StorageStub/StorageProviderBase API?

File: src/runtime/manifest.ts [hunk 3]
```diff
@@ -1052,11 +1061,13 @@ ${e.message}
         }
 
         if (targetParticle) {
-          let targetConnection;
-          if (connectionItem.target.param) {
-            targetConnection = targetParticle.connections[connectionItem.target.param];
+          let targetConnection: HandleConnection;
+
+          // TODO(lindner): replaced param with name since param is not defined, but name/particle are...
```
Comment (lindner) [inline src/runtime/manifest.ts]: Looks like this was last touched by @dstockwell back in 2017.  Checking the PEG grammar/ast types it appears that this was incorrect.  Thanks @csilvestrini

File: src/runtime/testing/stub-loader.ts [hunk 4]
```diff
@@ -8,44 +8,47 @@
...
+  constructor(fileMap: Dictionary<string>) {
     super();
     this._fileMap = fileMap;
     if (fileMap.hasOwnProperty('*')) {
       this._cannedResponse = fileMap['*'];
     }
   }
 
-  loadResource(path: string) {
+  // TODO(lindner): this should be async, but adding it causes a test to fail
```
Comment (lindner) [inline src/runtime/testing/stub-loader.ts]: Changing this to async causes the _ exposes a dereference API to particles_ test to fail.  Guessing it's some timing related issue.

File: src/planning/recipe-index.ts [hunk 5]
```diff
@@ -308,10 +315,16 @@ export class RecipeIndex {
...
    */
-  private _fatesAndDirectionsMatch(slotHandleConn, matchingHandleConn): boolean {
+  private _fatesAndDirectionsMatch(slotHandleConn: HandleConnection, matchingHandleConn: HandleConnection): boolean {
     const matchingHandle = matchingHandleConn.handle;
     const allMatchingHandleConns = matchingHandle ? matchingHandle.connections : [matchingHandleConn];
-    const matchingHandleConnsHasOutput = allMatchingHandleConns.find(conn => ['out', 'inout'].includes(conn.direction));
+    const matchingDirection: Direction[] = ['out', 'inout'];
+    const matchingHandleConnsHasOutput = allMatchingHandleConns.find(conn => conn.direction && matchingDirection.includes(conn.direction));
+
+    if (!slotHandleConn.handle) {
```
Comment (mmandlis) [inline src/planning/recipe-index.ts]: can this be an assert instead?

File: src/runtime/manifest.ts [hunk 6]
```diff
@@ -740,8 +749,7 @@ ${e.message}
     manifest._interfaces.push(ifaceInfo);
   }
 
-  // TODO(cypher): Remove loader dependency.
```
Comment (lindner) [inline src/runtime/manifest.ts]: It's gone @Cypher1 !

File: src/planning/recipe-index.ts [hunk 7]
```diff
@@ -308,10 +315,16 @@ export class RecipeIndex {
    * - `matchingHandleConn` - a handle connection of a particle, whose slot connection is explored
    *    as a potential match to a slot above.
    */
-  private _fatesAndDirectionsMatch(slotHandleConn, matchingHandleConn): boolean {
+  private _fatesAndDirectionsMatch(slotHandleConn: HandleConnection, matchingHandleConn: HandleConnection): boolean {
     const matchingHandle = matchingHandleConn.handle;
     const allMatchingHandleConns = matchingHandle ? matchingHandle.connections : [matchingHandleConn];
-    const matchingHandleConnsHasOutput = allMatchingHandleConns.find(conn => ['out', 'inout'].includes(conn.direction));
+    const matchingDirection: Direction[] = ['out', 'inout'];
+    const matchingHandleConnsHasOutput = allMatchingHandleConns.find(conn => conn.direction && matchingDirection.includes(conn.direction));
```
Comment (mmandlis) [inline src/planning/recipe-index.ts]: what's the reason for this change? i don't think that this is more readable.
also, HandleConnection has a `isOutput()` getter method, maybe it can be used here instead of string matching?

File: src/runtime/manifest.ts [hunk 8]
```diff
@@ -985,6 +993,7 @@ ${e.message}
         connection.tags = connectionItem.target ? connectionItem.target.tags : [];
         const direction = {'->': 'out', '<-': 'in', '=': 'inout', 'consume': '`consume', 'provide': '`provide'}[connectionItem.dir];
```
Comment (lindner) [inline src/runtime/manifest.ts]: question for @shans backtick-consume (`enum) is not included in the default Direction enum.  Details?

File: src/runtime/manifest.ts [hunk 9]
```diff
@@ -1241,12 +1255,16 @@ ${e.message}
...
+    // TODO(lindner): fromLiteral is not declared in the StorageProviderBase/StorageStub interface
+    // tslint:disable-next-line: no-any
+    await (store as any).fromLiteral({version, model});
   }
 
-  private static async _createStore(manifest, type: Type, name: string, id: string, tags: string[], item, originalId: string) {
+  private static async _createStore(manifest: Manifest, type: Type, name: string, id: string, tags: string[], item, originalId: string): Promise<StorageProviderBase | StorageStub> {
     const store = await manifest.createStore(type, name, id, tags);
-    store.source = item.source;
+    // TODO(lindner): Property 'source' does not exist on type 'StorageStub | StorageProviderBase'.
```
Comment (lindner) [inline src/runtime/manifest.ts]: Last touched by @wkorman - is store.source needed?  remove?

File: src/runtime/entity.ts [hunk 10]
```diff
@@ -72,7 +72,9 @@ class EntityInternals {
   }
 
   getId(): string {
-    assert(this.isIdentified(), 'getId() called on unidentified entity');
+    if (this.id === undefined) {
```
Comment (mmandlis) [inline src/runtime/entity.ts]: why is `throw new Error` better than an assert, 
and separately why did you not use `this.isIdentified()`

File: src/runtime/manifest.ts [hunk 11]
```diff
@@ -475,7 +484,7 @@ ${e.message}
       // Loading of imported manifests is triggered in parallel to avoid a serial loading
       // of resources over the network.
       await Promise.all(items.map(async item => {
-        if (item.kind === 'import') {
+        if (item.kind === 'import' && loader) {
```
Comment (mmandlis) [inline src/runtime/manifest.ts]: is `loader` ever null? maybe this should be an assert?

File: src/runtime/recipe/handle.ts [hunk 12]
```diff
@@ -180,8 +180,8 @@ export class Handle {
   get immediateValue() { return this._immediateValue; }
   set immediateValue(value: ParticleSpec) { this._immediateValue = value; }
 
-  static effectiveType(handleType: Type, connections: {type: Type, direction: string}[]) {
-    const variableMap = new Map();
+  static effectiveType(handleType: Type, connections: {type: Type|null|undefined, direction: string|undefined|null}[]) {
```
Comment (mmandlis) [inline src/runtime/recipe/handle.ts]: why is there both here `null` and `undefined`?

File: src/runtime/manifest.ts (hunk 3 above)
Comment (csilvestrini) [reply src/runtime/manifest.ts]: Good spotting! I don't know how you figured that out... this code is so hard to follow without types...

File: src/planning/recipe-index.ts (hunk 5 above)
Comment (lindner) [reply src/planning/recipe-index.ts]: Typescript will narrow a nullable/undefined type to non-nullable/defined when you use an if statement or other control flow.  Sadly assert() does not provide that.

The `preconditions.ts` class has checkDefined() and checkNotNull() and they will narrow nullness.  So far have only used those for testing.  WDYT?  Move those up a level and make them available for general use?

File: src/planning/recipe-index.ts (hunk 7 above)
Comment (lindner) [reply src/planning/recipe-index.ts]: `['out', 'inout']` is untyped in the find predicate, also conn.direction could be undefined.

Switched check for `isOutput()` which is much more readable.  Thanks!

File: src/runtime/entity.ts (hunk 10 above)
Comment (lindner) [reply src/runtime/entity.ts]: undefined narrowing again.  If you check this.id for undefined directly the compiler will  know that it is defined.  If you don't have this the null checking fails as this.id does not match the return type.

Another fix for this would look like:

```
 getId(): string {
    return checkDefined(this.id);
 }
```

This avoids the if-assert block and moves the check and null/defined-ness narrowing into checkDefined()

File: src/runtime/manifest.ts (hunk 11 above)
Comment (lindner) [reply src/runtime/manifest.ts]: Some tests are not supplying a loader.  It does seem reasonable to throw when you are attempting to process a manifest with an import and without a loader.

File: src/runtime/recipe/handle.ts (hunk 12 above)
Comment (lindner) [reply src/runtime/recipe/handle.ts]: handle-connection has it defined that way.  The type there can be undefined (never set) or explicitly set to null.  Annoying..  Would like to root out usages like this..

File: src/planning/planner.ts (hunk 1 above)
Comment (Cypher1) [reply src/planning/planner.ts]: Can you point me to the reuse?

Comment (Cypher1) [reply src/planning/planner.ts]: Ah, got it

File: src/runtime/manifest.ts (hunk 6 above)
Comment (Cypher1) [reply src/runtime/manifest.ts]: YAY! Thanks @lindner

File: src/runtime/manifest.ts (hunk 8 above)
Comment (Cypher1) [reply src/runtime/manifest.ts]: The \`consume and \`provide are still being incorporated. I'm working on it here: 
https://github.com/PolymerLabs/arcs/pull/3221

**Comment by lindner** [comment]
Maria is out until monday.  Going to move forward with this.  Happy to re-review any of these changes.

File: src/planning/recipe-index.ts (hunk 5 above)
Comment (mmandlis) [reply src/planning/recipe-index.ts]: I think it'll be nice to use them in the code as well - it's easy to mask bugs with the if statements, and having both - if and assert is too verbose.

**Comment by mmandlis** [comment]
belated LGTM, thanks Paul!
